micro-rdk/src/esp32/rpc_heap.rs

```diff
@@ -0,0 +1,52 @@
+use std::{ffi::c_void, mem::ManuallyDrop};
+
+use crate::{
+    common::grpc::{GrpcError, RpcAllocation},
+    esp32::esp_idf_svc::sys::{
+        heap_caps_free, heap_caps_malloc, MALLOC_CAP_8BIT, MALLOC_CAP_SPIRAM,
+    },
+};
+use bytes::{BufMut, Bytes, BytesMut};
+
+#[derive(Clone)]
+pub struct Esp32RpcHeapAllocation {
+    // We source the memory for this Vec from ESP-IDF's heap_caps_malloc
+    // because it will error when not enough space is available rather than panic.
+    // As a result we must wrap the Vec in a ManuallyDrop to prevent double-free, since
+    // the memory will not be managed by Rust's allocators
+    inner: ManuallyDrop<Vec<u8>>,
+    ptr: *mut u8,
+}
+
+impl RpcAllocation for Esp32RpcHeapAllocation {
+    fn get_allocation(size: usize) -> Result<Self, GrpcError> {
```

Comment:
I think these deserve their own error types and get propagated upward to `GrpcError`

Reply:
I'd prefer to save doing that until we have an implementation of the trait that would actually want to surface any other error